src/msac.rs

```diff
@@ -176,27 +182,59 @@ impl Default for MsacContext {
     }
 }
 
-impl MsacContext {
-    fn set_buf(&mut self, buf: &[u8]) {
-        let Range { start, end } = buf.as_ptr_range();
-        self.buf_pos = start;
-        self.buf_end = end;
+impl MsacAsmContext {
+    fn allow_update_cdf(&self) -> bool {
+        self.allow_update_cdf != 0
     }
 
-    fn with_buf(&mut self, mut f: impl FnMut(&[u8]) -> &[u8]) {
-        // # Safety
-        //
-        // [`Self::buf_pos`] and [`Self::buf_end`] are the start and end ptrs of the `buf` slice,
-        // and are only set in [`Self::set_buf`], which derives them from a valid slice.
+    fn with_buf<'a>(&mut self, mut f: impl FnMut(&'a [u8]) -> &'a [u8]) {
+        // SAFETY: [`Self::buf_pos`] and [`Self::buf_end`] are the start and end ptrs of the [`MsacContext::data`] slice.
+        // They are initially set in [`MsacContext::new`].
+        // [`Self::buf_pos`] may be modified in [`MsacContext::with_asm`] calls, but it remains valid.
+        // This [`Self::with_buf`] method itself is always called with [`MsacContext::with_asm`] higher on the call stack,
+        // so the owner of [`Self::buf_pos`], [`MsacContext::data`], is still alive and the lifetime is valid.
         let buf = unsafe {
             let len = self.buf_end.offset_from(self.buf_pos) as usize;
-            std::slice::from_raw_parts(self.buf_pos, len)
+            slice::from_raw_parts(self.buf_pos, len)
         };
-        self.set_buf(f(buf));
+        let Range { start, end } = f(buf).as_ptr_range();
+        self.buf_pos = start;
+        self.buf_end = end;
     }
+}
 
-    fn allow_update_cdf(&self) -> bool {
-        self.allow_update_cdf != 0
+#[derive(Default)]
+pub struct MsacContext {
+    asm: MsacAsmContext,
+    pub data: Option<CArc<[u8]>>,
+    pub data_offset: usize,
+}
+
+impl Deref for MsacContext {
+    type Target = MsacAsmContext;
+
+    fn deref(&self) -> &Self::Target {
+        &self.asm
+    }
+}
+
+impl DerefMut for MsacContext {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.asm
+    }
+}
+
+impl MsacContext {
+    fn with_asm<R>(&mut self, mut f: impl FnMut(&mut MsacAsmContext) -> R) -> R {
+        let data = self.data.as_ref().unwrap().as_ref();
+        // Hold this reference across `f`, which ensures [`MsacAsmContext::with_buf`] is safe.
+        let retval = f(&mut self.asm);
+        let data = data.as_ptr_range();
+        debug_assert!(data.end == self.buf_end);
+        debug_assert!(data.start <= self.buf_pos && data.end >= self.buf_pos);
+        // SAFETY: The asm `fn`s, `fn *_c`s, and `fn *_rust`s must keep `buf_pos` valid.
+        self.data_offset = unsafe { self.buf_pos.offset_from(data.start) as usize };
```

Comment:
The above asserts are checking that preconditions for `offset_from` are upheld right? We should mention that in the safety comment, even if they're debug-only asserts.

Reply:
Sure, I can.

Reply:
Do we need to compute and store `data_offset` every time here? Couldn't we compute it when needed?

Reply:
No more `data_offset` anymore.